FingerprintDialogKotlin/app/src/main/java/com/example/android/fingerprintdialog/MainActivity.kt

```diff
@@ -287,6 +280,53 @@ class MainActivity : AppCompatActivity(),
         return super.onOptionsItemSelected(item)
     }
 
+    private fun instanceOfBiometricPrompt(): BiometricPrompt {
+        val executor = ContextCompat.getMainExecutor(this)
+
+        val callback = object : BiometricPrompt.AuthenticationCallback() {
+            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
+                super.onAuthenticationError(errorCode, errString)
+                Log.d(TAG, "$errorCode :: $errString")
+                if(errorCode == BiometricPrompt.ERROR_NEGATIVE_BUTTON) {
+                    loginWithPassword() // because negative button says use application password
+                }
+            }
+
+            override fun onAuthenticationFailed() {
+                super.onAuthenticationFailed()
+                Log.d(TAG, "Authentication failed for an unknown reason")
+            }
+
+            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
+                super.onAuthenticationSucceeded(result)
+                Log.d(TAG, "Authentication was successful")
+                onPurchased(true,result.cryptoObject)
+            }
+        }
+
+        val biometricPrompt = BiometricPrompt(this, executor, callback)
+        return biometricPrompt
+    }
+
+    private fun getPromptInfo(): BiometricPrompt.PromptInfo {
+        val promptInfo = BiometricPrompt.PromptInfo.Builder()
+                .setTitle(getString(R.string.prompt_info_title))
+                .setSubtitle(getString(R.string.prompt_info_subtitle))
+                .setDescription(getString(R.string.prompt_info_description))
+                .setConfirmationRequired(false)
+                .setNegativeButtonText(getString(R.string.prompt_info_use_app_password))
+                // .setDeviceCredentialAllowed(true) // true for device credential option
```

Comment:
This comment could be clearer. Something like:
```
// .setDeviceCredentialAllowed(true) // Allow PIN/pattern/password authentication.
```

Also, just uncommenting that line won't work, since it's incompatible with `setNegativeButtonText` above.